test/hotspot/jtreg/compiler/lib/template_framework/Template.java

```diff
@@ -0,0 +1,599 @@
+ * the {@link Template}s provide <strong>hashtag replacements</strong> in the Strings: the {@link Template} argument
+ * names are captured, and the argument values automatically replace any {@code "#name"} in the Strings. See the
+ * different overloads of {@link #make} for examples. Additional hashtag replacements can be defined
+ * with {@link #let}.
+ *
+ * <p>
+ * When using nested {@link Template}s, there can be collisions with identifiers (e.g. variable names and method names).
+ * For this, {@link Template}s provide <strong>dollar replacements</strong>, which automaticall rename any
+ * {@code "$name"} in the String with a {@code "name_ID"}, where the {@code "ID"} is unique for every use of
+ * a {@link Template}. The dollar replacement can also be captured with {@link #$}, and passed to nested
+ * {@link Template}s, which allows sharing of these identifier names between {@link Template}s.
+ *
+ * <p>
+ * To render a {@link Template} to a {@link String}, one first has to apply the arguments (e.g. with
+ * {@link TwoArgs#withArgs}) and then the resulting {@link TemplateWithArgs} can either be used as a
+ * {@link Token} inside another {@link Template}, or rendered to a {@link String} with {@link TemplateWithArgs#render}.
+ *
+ * <p>
+ * A {@link TemplateWithArgs} can be used directly as a {@link Token} inside the {@link Template#body} to
+ * nest the {@link Template}s. Alternatively, code can be {@link Hook#insert}ed to where a {@link Hook}
+ * was {@link Hook#set} earlier (in some outer scope of the code). For example, while generating code in
+ * a method, one can reach out to the scope of the class, and insert a new field, or define a utility method.
+ *
+ * <p>
+ * A {@link TemplateBinding} allows the recurisve use of {@link Template}s. With the indirection of such a binding,
+ * a {@link Template} can reference itself. To ensure the termination of recursion, the templates are rendered
+ * with a certain amount of {@link #fuel}, which is decreased at each {@link Template} nesting by a certain amount
+ * (can be changed with {@link #setFuelCost}). Recursive templates are supposed to terminate once the {@link #fuel}
+ * is depleated (i.e. reaches zero).
+ *
+ * <p>
+ * Code generation often involves defining fields and variables, which are then available inside a defined
+ * scope, and can be sampled in any nested scope. To allow the use of names for multiple applications (e.g.
+ * fields, variables, methods, etc), we define a {@link Name}, which captures the {@link String} representation
+ * to be used in code, as well as its type and if it is mutable. One can add such a {@link Name} to the
+ * current code scope with {@link #addName}, and sample from the current or outer scopes with {@link #sampleName}.
+ * When generating code, one might want to create {@link Name}s (variables, fields, etc) in local scope, or
+ * in some outer scope with the use of {@link Hook}s.
```

Comment:
Maybe mention here again that all of the explained above can be found in tutorial like examples (I guess in `TestTutorial`)?. Because it was not that easy to grasp how these different options to create Templates now work in practice.

Reply:
Ok, fair. This is just a high level explanation. Especially Hooks and Names are also not the "starter features", I think. So it's ok if you have to go look at the examples or other uses, I think.

Reply:
I linked to the examples like `TestTutorial`.